vm/src/warn.rs

```diff
@@ -31,19 +62,130 @@ pub fn warn(
+            Err(vm.new_value_error(format!("_warnings.filters item {} isn't a 5-tuple", i)))
+        }?;
+
+        /* Python code: action, msg, cat, mod, ln = item */
+        let _action = tmp_item.get(0);
+        let _msg = tmp_item.get(1);
+        let _cat = tmp_item.get(2);
+        let _item_mod = tmp_item.get(3);
+        let _ln_obj = tmp_item.get(4);
+    }
+
+    Ok(())
+}
+
+fn already_warned(
+    registry: PyObjectRef,
+    key: PyObjectRef,
+    should_set: usize,
+    vm: &VirtualMachine,
+) -> bool {
+    let version_obj = registry.get_item("version", vm).ok();
+    let filters_version = vm
+        .ctx
+        .new_int(vm.state.warnings.filters_version)
+        .as_object()
+        .to_owned();
+
+    if version_obj.is_none()
+        || version_obj.as_ref().unwrap().try_int(vm).is_err()
+        || !version_obj.unwrap().is(&filters_version)
+    {
+        let registry = registry.dict();
+        registry.as_ref().map(|registry| {
+            registry.into_iter().collect_vec().clear();
+            registry
+        });
+
+        if let Some(registry) = registry {
+            if registry.set_item("version", filters_version, vm).is_err() {
+                return false;
+            }
+        }
+    } else {
+        let already_warned = registry.get_item(key.as_ref(), vm);
+        return already_warned.is_ok();
+    }
+
+    /* This warning wasn't found in the registry, set it. */
+    if should_set != 0 {
+        registry.set_item(key.as_ref(), vm.new_pyobj(""), vm).ok();
+    }
+
+    false
+}
+
+fn normalize_module(filename: PyStrRef, vm: &VirtualMachine) -> Option<PyObjectRef> {
+    let len = filename.char_len();
+
+    if len == 0 {
+        Some(vm.new_pyobj("<unknown>"))
+    } else if len >= 3 && filename.as_str().contains(".py") {
+        Some(vm.new_pyobj(filename.as_str().replace(".py", "")))
+    } else {
+        Some(filename.as_object().to_owned())
+    }
+}
+
 #[allow(clippy::too_many_arguments)]
 fn warn_explicit(
     category: Option<PyTypeRef>,
     message: PyStrRef,
-    _filename: PyStrRef,
-    _lineno: usize,
-    _module: PyObjectRef,
-    _registry: PyObjectRef,
+    filename: PyStrRef,
+    lineno: usize,
+    module: Option<PyObjectRef>,
+    registry: PyObjectRef,
     _source_line: Option<PyObjectRef>,
     _source: Option<PyObjectRef>,
     vm: &VirtualMachine,
 ) -> PyResult<()> {
-    // TODO: Implement correctly
+    if registry.clone().is_true(vm)?
+        && !registry.class().is(vm.ctx.types.dict_type)
+        && !registry.class().is(vm.ctx.types.none_type)
+    {
+        return Err(vm.new_type_error("'registry' must be a dict or None".to_string()));
+    }
+
+    // Normalize module.
+    let module = module.and(normalize_module(filename, vm));
+    let module = if let Some(module) = module {
+        module
+    } else {
+        return Ok(());
+    };
+
+    // Normalize message.
+    let text = message.as_str();
+    if text.is_empty() {
+        return Ok(());
+    }
```

Comment:
It seems text should not always be `str` type.
```suggestion
    let (category, text, message) = if message.is_instance(vm.ctx.exceptions.warning.as_object(), vm)? {
        (message.class().into_pyobject(vm), message.str(vm)?.into_pyobject(vm), message)
    } else {
        (category, message, vm.invoke(category.as_object(), (message, ))?)
    };
```
I am not sure that this is the best option. please take a look this code @youknowone

Reply:
the suggestion looks good.